dbms/src/DataStreams/MergingAndConvertingBlockInputStream.h

```diff
@@ -120,51 +123,58 @@ class MergingAndConvertingBlockInputStream : public IProfilingBlockInputStream
                 throw Exception("Unknown aggregated data variant.", ErrorCodes::UNKNOWN_AGGREGATED_DATA_VARIANT);
             }
 #undef M
-            return aggregator.prepareBlockAndFillSingleLevel(*first, final);
+            single_level_blocks = aggregator.prepareBlocksAndFillSingleLevel(*first, final);
+            return popBlocksListFront(single_level_blocks);
         }
         else
         {
+            if (Block out = popBlocksListFront(two_level_blocks))
+            {
+                return out;
+            }
             if (!parallel_merge_data)
             {
                 parallel_merge_data = std::make_unique<ParallelMergeData>(threads);
                 for (size_t i = 0; i < threads; ++i)
                     scheduleThreadForNextBucket();
             }
 
-            Block res;
-
             while (true)
             {
                 std::unique_lock lock(parallel_merge_data->mutex);
 
+                if (current_bucket_num >= NUM_BUCKETS)
+                {
+                    return {};
+                }
+
                 if (parallel_merge_data->exception)
                     std::rethrow_exception(parallel_merge_data->exception);
 
                 auto it = parallel_merge_data->ready_blocks.find(current_bucket_num);
                 if (it != parallel_merge_data->ready_blocks.end())
                 {
-                    ++current_bucket_num;
                     scheduleThreadForNextBucket();
+                    current_bucket_num++;
 
-                    if (it->second)
+                    if (!it->second.empty())
                     {
-                        res.swap(it->second);
-                        break;
+                        two_level_blocks.splice(two_level_blocks.end(), std::move(it->second), it->second.begin(), it->second.end());
+                        return popBlocksListFront(two_level_blocks);
```

Comment:
Here we still need to check if retuned `block` is empty?

Reply:
> Here we still need to check if retuned `block` is empty?

done.